edb/schema/pointers.py

```diff
@@ -228,21 +229,128 @@ def merge_target(
 
         if target is None:
             target = base_target
+            current_source = base.get_source(schema)
         else:
-            schema, target = Pointer.merge_targets(
-                schema, ptr, target, base_target, allow_contravariant=True)
+            assert current_source is not None
+            source = base.get_source(schema)
+            assert source is not None
+            schema, target = _merge_types(
+                schema,
+                ptr,
+                target,
+                base_target,
+                t1_source=current_source,
+                t2_source=source,
+                allow_contravariant=True,
+            )
 
     if not ignore_local:
         local_target = ptr.get_target(schema)
         if target is None:
             target = local_target
         elif local_target is not None:
-            schema, target = Pointer.merge_targets(
-                schema, ptr, target, local_target)
+            assert current_source is not None
+            schema, target = _merge_types(
+                schema,
+                ptr,
+                target,
+                local_target,
+                t1_source=current_source,
+                t2_source=None,
+            )
 
     return target
 
 
+def _merge_types(
+    schema: s_schema.Schema,
+    ptr: Pointer,
+    t1: s_types.Type,
+    t2: s_types.Type,
+    *,
+    t1_source: so.Object,
+    t2_source: Optional[so.Object],
+    allow_contravariant: bool = False,
+) -> Tuple[s_schema.Schema, Optional[s_types.Type]]:
+    if t1 == t2:
+        return schema, t1
+
+    # When two pointers are merged, check target compatibility
+    # and return a target that satisfies both specified targets.
+
+    if (isinstance(t1, s_abc.ScalarType) !=
+            isinstance(t2, s_abc.ScalarType)):
+        # Mixing a property with a link.
+        vnp = ptr.get_verbosename(schema, with_parent=True)
+        vn = ptr.get_verbosename(schema)
+        t1_vn = t1.get_verbosename(schema)
+        t2_vn = t2.get_verbosename(schema)
+        t1_cls = 'property' if isinstance(t1, s_abc.ScalarType) else 'link'
+        t2_cls = 'property' if isinstance(t2, s_abc.ScalarType) else 'link'
+
+        t1_source_vn = t1_source.get_verbosename(schema, with_parent=True)
+        if t2_source is None:
+            raise errors.SchemaError(
+                f'cannot redefine {vnp} as {t2_vn}',
+                details=(
+                    f'{vn} is defined as a {t1_cls} to {t1_vn} in'
+                    f' parent {t1_source_vn}'
```

Comment:
```suggestion
                    f' parent type {t1_source_vn}'
```

Reply:
The result of `get_verbosename()` already includes the word "type".